src/client/texturepaths.cpp

```diff
@@ -0,0 +1,113 @@
+/*
+Minetest
+Copyright (C) 2010-2013 celeron55, Perttu Ahola <celeron55@gmail.com>
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "texturepaths.h"
+
+#include "util/container.h"
+#include "settings.h"
+#include "filesys.h"
+#include "porting.h"
+#include <vector>
+
+// A cache from texture name to texture path
+static MutexedMap<std::string, std::string> g_texturename_to_path_cache;
+
+void clearTextureNameCache()
+{
+	g_texturename_to_path_cache.clear();
+}
+
+// Find out the full path of an image by trying different filename extensions.
+// If failed, return "".
+std::string getImagePath(std::string_view path)
+{
+	// A NULL-ended list of possible image extensions
+	// (In newer C++ versions a fixed size array and ranges::subrange could be used
+	// or just modernise removeStringEnd.)
+	std::vector<const char*> extensions{".png", ".jpg", ".bmp", ".tga", nullptr};
+
+	// Remove present extension
+	std::string_view stripped_path = removeStringEnd(path, extensions.data());
+	// If there is no known extension, assume it has been omitted.
+	if (stripped_path.empty())
+		stripped_path = path;
+
+	extensions.pop_back();
+	for (const char *ext : extensions) {
+		std::string extended_path{stripped_path};
+		extended_path.append(ext);
+		if (fs::PathExists(extended_path))
+			return extended_path;
+	}
```

Comment:
surely this would have been simpler?
```suggestion
	const char *extensions[] = {".png", ".jpg", ".bmp", ".tga", nullptr};

	// Remove present extension
	std::string_view stripped_path = removeStringEnd(path, extensions);
	// If there is no known extension, assume it has been omitted.
	if (stripped_path.empty())
		stripped_path = path;

	for (const char *ext : extensions) {
		if (!ext)
			break;
		std::string extended_path(stripped_path);
		extended_path.append(ext);
		if (fs::PathExists(extended_path))
			return extended_path;
	}
```

Reply:
Unnecessarily calling `if (!ext)` 5 times for every path does not look very nice imo. (even though it will probably be optimized away)

Ideally, `removeStringEnd` should be changed to take a vector as parameter instead of a NULL terminated C array, I think.

Do you want me to apply your suggestion, or can you live with my code?